File: frontend/src/lib/api/practice.ts
/**
 * API client for Practice Sessions
 * Similar to diagnostic but focuses on a single OA with adaptive difficulty
 */

import { auth } from '$lib/stores/auth.svelte';

export interface PracticeSession {
  id: number;
  user_id: number;
  oa_id: number;
  oa_bloom_objective_id: number;
  bloom_level_inicial: number;
  bloom_level_final?: number;
  numero_preguntas: number;
  preguntas_respondidas: number;
  preguntas_correctas: number;
  estado: 'en_progreso' | 'completado';
  estrategia: any;
  resultado?: any;
  started_at: string;
  completed_at?: string;
  oa?: any;
}

export interface PracticeQuestion {
  id: number;
  oa_bloom_objective_id: number;
  tipo: string;
  question_data: any;
  question_number: number;
  total_questions: number;
  current_bloom_level: number;
}

export interface PracticeResult {
  session: PracticeSession;
  bloom_level_inicial: number;
  bloom_level_final: number;
  cambio_nivel: number;
  resultado: {
    porcentaje_aciertos: number;
    preguntas_totales: number;
    preguntas_correctas: number;
    aciertos_por_nivel: Record<number, number>;
    fallos_por_nivel: Record<number, number>;
    patron_respuestas: string[];
  };
}

function getAuthHeaders() {
  const token = auth.token;
  return {
    'Content-Type': 'application/json',
    'Authorization': token ? `Bearer ${token}` : ''
  };
}

/**
 * Start a new practice session for a specific OA
 */
export async function startPracticeSession(
  oaId: number,
  oaBloomObjectiveId: number,
  numeroPreguntas: number = 10
): Promise<{ success: boolean; session?: PracticeSession; error?: string }> {
  try {
    const response = await fetch('/api/practice-sessions', {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        oa_id: oaId,
        oa_bloom_objective_id: oaBloomObjectiveId,
        numero_preguntas: numeroPreguntas
      })
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const session = await response.json();
    return { success: true, session };
  } catch (error) {
    console.error('Error starting practice session:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
}

/**
 * Get list of practice sessions with optional filters
 */
export async function getPracticeSessions(filters?: {
  oa_id?: number;
  estado?: 'en_progreso' | 'completado';
}): Promise<{ success: boolean; sessions?: PracticeSession[]; error?: string }> {
  try {
    const params = new URLSearchParams();
    if (filters?.oa_id) params.append('oa_id', filters.oa_id.toString());
    if (filters?.estado) params.append('estado', filters.estado);

    const response = await fetch(`/api/practice-sessions?${params.toString()}`, {
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const sessions = await response.json();
    return { success: true, sessions };
  } catch (error) {
    console.error('Error fetching practice sessions:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
}

/**
 * Get next question in practice session with adaptive difficulty
 */
export async function getNextPracticeQuestion(
  sessionId: number
): Promise<{ success: boolean; question?: PracticeQuestion; error?: string }> {
  try {
    const response = await fetch(`/api/practice-sessions/${sessionId}/next-question`, {
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const question = await response.json();
    return { success: true, question };
  } catch (error) {
    console.error('Error fetching next question:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
}

/**
 * Submit an answer to a practice question
 */
export async function submitPracticeAnswer(
  sessionId: number,
  questionId: number,
  userAnswer: any,
  tiempoSegundos?: number
): Promise<{
  success: boolean;
  is_correct?: boolean;
  score?: number;
  new_bloom_level?: number;
  is_complete?: boolean;
  error?: string;
}> {
  try {
    const response = await fetch(`/api/practice-sessions/${sessionId}/answer`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        question_id: questionId,
        user_answer: userAnswer,
        tiempo_segundos: tiempoSegundos
      })
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const result = await response.json();
    return { success: true, ...result };
  } catch (error) {
    console.error('Error submitting answer:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
}

/**
 * Complete practice session and get final results
 */
export async function completePracticeSession(
  sessionId: number
): Promise<{ success: boolean; result?: PracticeResult; error?: string }> {
  try {
    const response = await fetch(`/api/practice-sessions/${sessionId}/complete`, {
      method: 'POST',
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const result = await response.json();
    return { success: true, result };
  } catch (error) {
    console.error('Error completing practice session:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
}

File: backend/internal/models/practice.go
package models

import (
	"time"

	"gorm.io/datatypes"
)

// PracticeSession represents a practice session for a specific OA
type PracticeSession struct {
	ID                  uint           `json:"id" gorm:"primaryKey"`
	UserID              uint           `json:"user_id" gorm:"not null;index"`
	OAID                uint           `json:"oa_id" gorm:"not null;index"`
	OABloomObjectiveID  uint           `json:"oa_bloom_objective_id" gorm:"not null"` // Specific OA+Bloom level
	BloomLevelInicial   int            `json:"bloom_level_inicial"`                  // Starting level
	BloomLevelFinal     *int           `json:"bloom_level_final"`                    // Final level after practice
	NumeroPreguntas     int            `json:"numero_preguntas" gorm:"default:10"`   // Number of questions (default 10)
	PreguntasRespondidas int           `json:"preguntas_respondidas" gorm:"default:0"`
	PreguntasCorrectas  int            `json:"preguntas_correctas" gorm:"default:0"`
	Estado              string         `json:"estado" gorm:"type:varchar(20);default:'en_progreso';index"` // en_progreso, completado
	Estrategia          datatypes.JSON `json:"estrategia" gorm:"type:jsonb"`         // Adaptive strategy data
	Resultado           datatypes.JSON `json:"resultado" gorm:"type:jsonb"`          // Final results and analysis
	StartedAt           time.Time      `json:"started_at"`
	CompletedAt         *time.Time     `json:"completed_at"`
	CreatedAt           time.Time      `json:"created_at"`
	UpdatedAt           time.Time      `json:"updated_at"`

	// Relations
	User             User             `json:"user,omitempty" gorm:"foreignKey:UserID"`
	OA               ObjetivoAprendizaje `json:"oa,omitempty" gorm:"foreignKey:OAID"`
	OABloomObjective OABloomObjective `json:"oa_bloom_objective,omitempty" gorm:"foreignKey:OABloomObjectiveID"`
	Answers          []PracticeAnswer `json:"answers,omitempty" gorm:"foreignKey:SessionID"`
}

// PracticeAnswer represents a single answer in a practice session
type PracticeAnswer struct {
	ID                 uint           `json:"id" gorm:"primaryKey"`
	SessionID          uint           `json:"session_id" gorm:"not null;index"`
	QuestionID         uint           `json:"question_id" gorm:"not null"`
	BloomLevelID       uint           `json:"bloom_level_id" gorm:"not null"` // Difficulty level of question
	UserAnswer         datatypes.JSON `json:"user_answer" gorm:"type:jsonb"`
	IsCorrect          *bool          `json:"is_correct"`
	Score              *float64       `json:"score"`              // Partial credit (0-1)
	TiempoSegundos     *int           `json:"tiempo_segundos"`
	CreatedAt          time.Time      `json:"created_at"`

	// Relations
	Session  PracticeSession `json:"session,omitempty" gorm:"foreignKey:SessionID"`
	Question Question        `json:"question,omitempty" gorm:"foreignKey:QuestionID"`
}

// PracticeStrategy represents the adaptive strategy used during practice
type PracticeStrategy struct {
	NivelBloomActual     int      `json:"nivel_bloom_actual"`
	AciertosConsecutivos int      `json:"aciertos_consecutivos"`
	FallosConsecutivos   int      `json:"fallos_consecutivos"`
	AciertosPorNivel     map[int]int `json:"aciertos_por_nivel"`
	FallosPorNivel       map[int]int `json:"fallos_por_nivel"`
	PatronRespuestas     []string `json:"patron_respuestas"` // "C" = correct, "I" = incorrect
}

func (PracticeSession) TableName() string {
	return "practice_sessions"
}

func (PracticeAnswer) TableName() string {
	return "practice_answers"
}
